google/cloud/spanner_dbapi/client_side_statement_executor.py

```diff
@@ -0,0 +1,27 @@
+# Copyright 20203 Google LLC All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from google.cloud.spanner_dbapi.parsed_statement import (
+    ParsedStatement,
+    ClientSideStatementType,
+)
+
+
+def execute(connection, parsed_statement: ParsedStatement):
```

Comment:
IMO `single leading underscore` means that the function should not be accessed outside the module, which we need to access in this case. Also Intellij is giving the warning if I add an underscore to the function.

As per https://peps.python.org/pep-0008/#public-and-internal-interfaces we dont have this function listed here https://github.com/googleapis/python-spanner/blob/07fbc45156a1b42a5e61c9c4b09923f239729aa8/google/cloud/spanner_dbapi/__init__.py#L61 so it should not be considered a public interface but its just a convention and like everything else in python the function can be accessed as a public interface. Should we have something like an internalapi annotation from Java?

@aseering if you have any suggestions

Reply:
What do you mean with 'which we need to access in this case'? Do you mean that we will be accessing this from our Django/SQLAlchemy implementations? Or do you mean from some other module in the dbapi driver?

I would consider our Django and SQLAlchemy implementations 'internal' in this case as well, although maybe technically they are not. The difference here is that:
1. If we publish this function as public to outside users, then we cannot change the signature or behavior of it later.
2. If we mark it as internal, then we are free to change it if necessary, and as we are in control of for example our Django and SQLAlchemy implementations, so we could easily coordinate any changes to this function with the usage there.

From https://peps.python.org/pep-0008/#public-and-internal-interfaces:

> Even with \_\_all\_\_ set appropriately, internal interfaces (packages, modules, classes, functions, attributes or other names) should still be prefixed with a single leading underscore.

Reply:
At discussed offline added a comment stating that its is an internal method that can make backwards-incompatible changes and not using underscores.

Reply:
> IMO `single leading underscore` means that the function should not be accessed outside the module, which we need to access in this case.

In my experience, leading underscore simply means "if you're not the one who wrote this function, it's not for you to use."  Scoping to any particular namespace is not expected.  So I would not expect this to need to be "public" for that reason.

Surfacing a standard interface between components is a good practice to strive for, though.  Is this the interface that we want to standardize on?

Reply:
Thanks Adam

> Surfacing a standard interface between components is a good practice to strive for, though. Is this the interface that we want to standardize on?

I didn't understand this

Reply:
I guess what I'm trying to say is, if we want `def execute(...)` to be public, we should spend a bit of time thinking about whether it's the right interface for the job.  Are we ok with supporting this function with this name and this signature in this file, ~forever?  If a bunch of customers find it and decide to use it, are we ok with that?  Or would we wish that we'd spent a bit more time thinking and designing because there's clearly a different signature that would be better here?

(If the function is marked with an underscore, to some extent that communicates that we haven't spent that time reviewing and committing to this as a public interface, so we might discover a mistake and change it in the future.)

Reply:
Thanks for the explanation. I had already commented on the function  that `its is an internal method that can make backwards-incompatible changes` as we were confused about using underscore. But now that you have explained it, I will refactor this function (and few others) to use underscore in a separate PR